ext/drbbdup/drbbdup.c

```diff
@@ -703,13 +757,112 @@ drbbdup_insert_dispatch(void *drcontext, instrlist_t *bb, instr_t *where,
     drbbdup_insert_landing_restoration(drcontext, bb, where, manager);
 }
 
+/* Returns whether or not additional cases should be handled poised by copy limit
+ * defined by the user.
+ */
+static bool
+drbbdup_do_dynamic_handling(drbbdup_manager_t *manager)
+{
+    drbbdup_case_t *drbbdup_case;
+    int i;
+    for (i = 0; i < opts.dup_limit; i++) {
+        drbbdup_case = &(manager->cases[i]);
+        /* Search for empty undefined slot. */
+        if (!drbbdup_case->is_defined)
+            return true;
+    }
+
+    return false;
+}
+
+/* Insert trigger for dynamic case handling. */
+static void
+drbbdup_insert_dynamic_handling(void *drcontext, app_pc translation_pc, void *tag,
+                                instrlist_t *bb, instr_t *where,
+                                drbbdup_manager_t *manager)
+{
+    instr_t *instr;
+    opnd_t opnd;
+
+    instr_t *done_label = INSTR_CREATE_label(drcontext);
+    opnd_t mask_opnd = opnd_create_reg(DRBBDUP_SCRATCH_REG);
+
+    /* Check whether case limit  has not been reached. */
+    if (drbbdup_do_dynamic_handling(manager)) {
+        drbbdup_case_t *default_info = &(manager->default_case);
+        ASSERT(default_info->is_defined, "default case must be defined");
+
+        /* Jump if runtime encoding matches default encoding.
+         * Unknown encoding encountered upon fall-through.
+         */
+        opnd = opnd_create_immed_uint((uintptr_t)default_info->encoding, OPSZ_PTR);
+        instr = INSTR_CREATE_cmp(drcontext, mask_opnd, opnd);
+        instrlist_meta_preinsert(bb, where, instr);
+
+        instr = INSTR_CREATE_jcc(drcontext, OP_jz, opnd_create_instr(done_label));
+        instrlist_meta_preinsert(bb, where, instr);
+
+        /* Don't bother insertion if threshold limit is zero. */
+        if (opts.hit_threshold > 0) {
+            /* Update hit count and check whether threshold is reached. */
+            opnd_t hit_table_opnd = drbbdup_get_tls_raw_slot_opnd(DRBBDUP_HIT_TABLE_SLOT);
+
+            /* Load the hit counter table. */
+            instr = INSTR_CREATE_mov_ld(drcontext, mask_opnd, hit_table_opnd);
```

Comment:
suggest: try to use XINST_CREATE wherever it's simple to do so.  It will make enabling for non-x86 that much easier in the future.

Reply:
I'll punt this for now but will file an issue.